snapcraft/internal/lxd.py

```diff
@@ -246,13 +253,27 @@ def _inject_snap(self, name):
             shutil.copyfile(installed, filepath)
         container_filename = os.path.join(os.sep, 'run', filename)
         self._push_file(filepath, container_filename)
-        logger.info('Installing {}'.format(container_filename))
-        cmd = ['snap', 'install', container_filename]
-        if rev.startswith('x'):
-            cmd.append('--dangerous')
+        self._install_snap(container_filename,
+                           is_dangerous=rev.startswith('x'),
+                           is_classic=is_classic)
+
+    def _install_snap(self, name, channel=None,
+                      is_dangerous=False,
+                      is_classic=False):
+        logger.info('Installing {}'.format(name))
+        # Install if needed
+        args = []
+        if channel:
+            args.append('--channel')
+            args.append(channel)
+        if is_dangerous:
+            args.append('--dangerous')
         if is_classic:
-            cmd.append('--classic')
-        self._container_run(cmd)
+            args.append('--classic')
+        self._container_run(['snap', 'install', name] + args)
+        if channel:
+            # Switch channel if needed
+            self._container_run(['snap', 'refresh', name] + args)
```

Comment:
I'm curious why we're doing this in a step separated from the install. I guess I have two questions:

1. Why not just use `--channel=<cannel>` in the install?
2. What do we expect to happen if we refresh to (or install with) a channel when installing with `--dangerous`?

Reply:
I updated the comments in an attempt to clarify: `snap install` will do nothing if the snap's already installed, even if the channel is different. So the refresh is needed to switch it - it won't do anything if the channel is the same. That's just how the command works...
There is no case where `--dangerous` is used with a channel - if it's a different arch and we `snap install` by name plus channel, there's no reasonable way to push an x-revisioned snap then, so we use the regular one. Does this make sense to you?

Reply:
Ah, I see-- I actually missed that you were also passing the channel to install. Carry on.